packages/core/status/core-status-server-internal/src/log_plugins_status.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Observable } from 'rxjs';
+import { pairwise, takeUntil, map, filter, startWith } from 'rxjs/operators';
+import type { PluginName } from '@kbn/core-base-common';
+import { ServiceStatusLevels } from '@kbn/core-status-common';
+import type { PluginStatus } from './plugins_status';
+
+export type PluginLevel = 'unavailable' | 'critical' | 'degraded' | 'available';
+export type PluginStatusWithName = PluginStatus & {
+  pluginName: PluginName;
+};
+export type PluginsByStatus = Record<PluginLevel, PluginStatusWithName[]>;
+
+export interface PluginStatusChanges {
+  status: PluginsByStatus;
+  updates: PluginsByStatus;
+  updated: number;
+  total: number;
+}
+
+export const getPluginsStatusChanges = (
+  plugins$: Observable<Record<PluginName, PluginStatus>>,
+  stop$: Observable<void>
+): Observable<PluginStatusChanges> => {
+  return plugins$.pipe(
+    startWith({}), // consider all plugins unavailable by default
+    takeUntil(stop$),
+    pairwise(),
+    map(([oldStatus, newStatus]) => getPluginsStatusDiff(oldStatus, newStatus)),
+    filter((statusChanges) => statusChanges.updated > 0)
+  );
+};
+
+const getPluginsStatusDiff = (
+  previous: Record<PluginName, PluginStatus>,
+  next: Record<PluginName, PluginStatus>
+): PluginStatusChanges => {
+  let updated: number = 0;
+  const status: PluginsByStatus = {
+    unavailable: [],
+    critical: [],
+    degraded: [],
+    available: [],
+  };
+  const updates: PluginsByStatus = {
+    unavailable: [],
+    critical: [],
+    degraded: [],
+    available: [],
+  };
+
+  Object.entries(next).forEach(([pluginName, pluginStatus]) => {
+    const currentLevel = pluginStatus.level;
+    status[currentLevel.toString()].push({ pluginName, ...pluginStatus });
+    const previousLevel = previous[pluginName]?.level ?? ServiceStatusLevels.unavailable;
+    if (previousLevel !== pluginStatus.level) {
+      // this plugin status has changed
+      updates[currentLevel.toString()].push({
+        pluginName,
+        ...pluginStatus,
+      } as PluginStatusWithName);
```

Comment:
NIT: we're constructing the exact same spreaded structure twice (`{ pluginName, ...pluginStatus }`). It can be reused between line 6 and 65 instead.